Custom Filters [FrontEnd]

nocommentcode opened the pull request:
As recommended by @Bruception , This PR is for the FrontEnd and will ovveride #3084 
NOTE: I will rebase this pr onto main once #3105 is merged, so ignore the backend changes in the pr, they are from the previous pr

### Description

"As a user i would like to be able to save filter configurations to be able to generate mutiple reports"

This PR allows users to create custom filters, which are just `ResultFilter` instances.
Previously the front end would store the `ResultFilter` in the local storage on the browser, however this means only 1 filter can be saved, and it is constantly overriden.
This PR now stores these custom filters in the backend database.

### UPDATED UI AND FLOW:

#### Updated flow:
- When the user clicks on the preset button, the filter preset section opens up
- The user can click on “save filter as preset” which will prompt them to select a name
- When the user clicks on an existing preset, the preset button is highlighted and the filters are set to the preset
- When the user changes the filters in any way (time frame, language, mode etc) the custom filter button is deselected, as the applied filters are no longer matching the preset 
- When a user selects a preset and its button is highlighted, the “delete preset” button appears
- When the user presses on the delete button, the preset is deleted


https://user-images.githubusercontent.com/26475724/173274961-a31f2989-cb9a-452c-a75f-949705ad7d85.mp4

Bruception commented:
So in terms of styling, I don't think the popup is acceptable to our standards. I'm not sure if @Miodec will touch it up. But if he does you will have to wait for him

nocommentcode commented:
> So in terms of styling, I don't think it is acceptable to our standards. I'm not sure if @Miodec will touch it up. But if he does you will have to wait for him to do so.

I have actually updated the styling, so it no longer looks like the video, I standardised it to be like the other components on the page, do have a look at the updated style and let me know if it still needs work

Bruception commented:
> > So in terms of styling, I don't think it is acceptable to our standards. I'm not sure if @Miodec will touch it up. But if he does you will have to wait for him to do so.
> 
> I have actually updated the styling, so it no longer looks like the video, I standardised it to be like the other components on the page, do have a look at the updated style and let me know if it still needs work

Sounds good. One think that completely slipped my mind is that we enforce naming restrictions. So names of filters can only have certain characters (just like tags or presets). Could you create a new PR that handles that appropriately?

Bruception commented:
Also if possible. It would be easier to discuss this feature in the development discord channel.

nocommentcode commented:
> Also if possible. It would be easier to discuss this feature in the development discord channel.

It wont let me join the discord server, it says the invitation link is invalid

nocommentcode commented:
> > > So in terms of styling, I don't think it is acceptable to our standards. I'm not sure if @Miodec will touch it up. But if he does you will have to wait for him to do so.
> > 
> > 
> > I have actually updated the styling, so it no longer looks like the video, I standardised it to be like the other components on the page, do have a look at the updated style and let me know if it still needs work
> 
> Sounds good. One think that completely slipped my mind is that we enforce naming restrictions (not sure why). So names of filters can only have certain characters (just like tags or presets). Could you create a new PR that handles that appropriately (backend validation)?

see #3110

Miodec commented:
From the UX and UI standpoint, i would prefer for it to work as presets (and call it filter presets). Currently its pretty weird, the way clicking custom just opens up the new menu, the add and remove buttons don't line up with the rest of the website, clicking anything else removes the menu. What I would change:

 - Replace custom button with a save icon, add a balloon that says "save filters as preset".
 - Dont keep the presets inside the advanced filters, keep them above the filters (and hide them if there are no presets saved)

Bruception commented:
> From the UX and UI standpoint, i would prefer for it to work as presets (and call it filter presets). Currently its pretty weird, the way clicking custom just opens up the new menu, the add and remove buttons don't line up with the rest of the website, clicking anything else removes the menu. What I would change:
> 
> * Replace custom button with a save icon, add a balloon that says "save filters as preset".
> * Dont keep the presets inside the advanced filters, keep them above the filters (and hide them if there are no presets saved)

I have renamed our data model:
https://github.com/monkeytypegame/monkeytype/pull/3115

@nocommentcode 
You will have to make the necessary changes in the frontend

nocommentcode commented:
@Miodec @Bruception 
I have updated the front end to now use the "filter preset" terminology
I also updated the UI, although not exactly how you explained it @Miodec 

Here is the flow at the moment, let me know what you think and if it needs changes:
- When the user clicks on the preset button, the filter preset section opens up
- The user can click on “save filter as preset” which will prompt them to select a name
- When the user clicks on an existing preset, the preset button is highlighted and the filters are set to the preset
- When the user changes the filters in any way (time frame, language, mode etc) the custom filter button is deselected, as the applied filters are no longer matching the preset 
- When a user selects a preset and its button is highlighted, the “delete preset” button appears
- When the user presses on the delete button, the preset is deleted


https://user-images.githubusercontent.com/26475724/173274961-a31f2989-cb9a-452c-a75f-949705ad7d85.mp4

Miodec commented:
Still some changes please:
 - move presets above the filters section
 - replace the "presets" button with a save icon button. this makes presets always visible (and moves the save button)
 - hide presets if no presets are saved
 - add a delete button for each preset, just like config presets or tags

nocommentcode commented:
> Still some changes please:
> 
> * move presets above the filters section
> * replace the "presets" button with a save icon button. this makes presets always visible (and moves the save button)
> * hide presets if no presets are saved
> * add a delete button for each preset, just like config presets or tags

Ok all done :) : 

https://user-images.githubusercontent.com/26475724/173497348-b00f4e05-3845-4381-83ec-89e5ebf08fee.mp4
